### PR TITLE
benches: add hierarchy bench

### PR DESCRIPTION
**Pull request by Hijiao**
## What have you changed? (mandatory)
Bench TiKV layered, here are the target methods:
+ engine::{get, put, snapshot}
+ mvcc::{prewrite, commit, seek_write, load_lock}
+ txn::{prewrite, commit}
+ storage::{prewrite, commit, raw_get}

## What are the type of the changes? (mandatory)

- Improvement (non-breaking change which is an improvement to an existing feature)

## How has this PR been tested? (mandatory)
passed bench test and works well with Schrodinger.

## Does this PR affect documentation (docs) update? (mandatory)
No

## Does this PR affect tidb-ansible update? (mandatory)
No

### CODE REVIEW

**Comment by Hijiao** [comment]
@hicqu @breeswish  PLAT thanks

File: benches/hierarchy/mvcc/mod.rs [hunk 1]
```diff
@@ -0,0 +1,162 @@
...
+            for key in &test_keys {
+                let mut reader = MvccReader::new(
+                    snapshot.clone(),
+                    None,
+                    false,
+                    None,
+                    None,
+                    ctx.isolation_level,
+                );
+                black_box(reader.seek_write(&key, u64::max_value()).unwrap());
```
Comment (breeswish) [inline benches/hierarchy/mvcc/mod.rs]: seek_write is an interface to be removed. I think it is not necessary to bench it here.

Comment (Hijiao) [reply benches/hierarchy/mvcc/mod.rs]: Ok, I will remove it later.

File: benches/hierarchy/txn/mod.rs [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+                    .generate(DEFAULT_ITERATIONS)
+                    .iter()
+                    .map(|(k, v)| (Mutation::Put((Key::from_raw(&k), v.clone())), k.clone()))
+                    .collect();
+            (mutations, &option)
+        },
+        |(mutations, option)| {
+            for (mutation, primary) in mutations {
+                let snapshot = engine.snapshot(&ctx).unwrap();
+                let mut txn = MvccTxn::new(snapshot, 1, false).unwrap();
```
Comment (zhangjinpeng1987) [inline benches/hierarchy/txn/mod.rs]: Please set `fill_cache` true.

File: benches/hierarchy/txn/mod.rs [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+}
+
+fn txn_commit<E: Engine, F: EngineFactory<E>>(b: &mut Bencher, config: &BenchConfig<F>) {
+    let engine = config.engine_factory.build();
+    let ctx = Context::new();
+    let option = Options::default();
+    b.iter_with_setup(
+        || {
+            let snapshot = engine.snapshot(&ctx).unwrap();
+            let mut txn = MvccTxn::new(snapshot, 1, false).unwrap();
```
Comment (zhangjinpeng1987) [inline benches/hierarchy/txn/mod.rs]: Ditto

File: benches/hierarchy/txn/mod.rs [hunk 4]
```diff
@@ -0,0 +1,86 @@
...
+            }
+            let modifies = txn.into_modifies();
+            let _ = engine.write(&ctx, modifies);
+            let keys: Vec<Key> = kvs.iter().map(|(k, _)| Key::from_raw(&k)).collect();
+            keys
+        },
+        |keys| {
+            for key in keys {
+                let snapshot = engine.snapshot(&ctx).unwrap();
+                let mut txn = MvccTxn::new(snapshot, 1, false).unwrap();
```
Comment (zhangjinpeng1987) [inline benches/hierarchy/txn/mod.rs]: Ditto

**Comment by zhangjinpeng1987** [comment]
Please provide some benchmark result after enable fill_cache.

**Comment by breeswish** [comment]
Interesting! BTW why does the benchmark result become better recently? We didn't change MVCC.

**Comment by Hijiao** [comment]
Thanks, @breeswish  this benchmark runs on 5 different machines by now. Hardware makes the curve, not our code.

**Pull request merged**